// eslint/lib/rules/no-delete-var.js
/*
 * © 2022-2026 Ashraf Morningstar
 * GitHub: https://github.com/AshrafMorningstar
 *
 * This project is a personal recreation of existing projects, developed by Ashraf Morningstar 
 * for learning and skill development. Original project concepts remains the intellectual 
 * property of their respective creators.
 */
/**
 * @fileoverview Rule to flag when deleting variables
 * @author Ilya Volodin
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

/** @type {import('../types').Rule.RuleModule} */
module.exports = {
	meta: {
		type: "suggestion",

		docs: {
			description: "Disallow deleting variables",
			recommended: true,
			url: "https://eslint.org/docs/latest/rules/no-delete-var",
		},

		schema: [],

		messages: {
			unexpected: "Variables should not be deleted.",
		},
	},

	create(context) {
		return {
			UnaryExpression(node) {
				if (
					node.operator === "delete" &&
					node.argument.type === "Identifier"
				) {
					context.report({ node, messageId: "unexpected" });
				}
			},
		};
	},
};
