// eslint/lib/rules/no-with.js
/*
 * © 2022-2026 Ashraf Morningstar
 * GitHub: https://github.com/AshrafMorningstar
 *
 * This project is a personal recreation of existing projects, developed by Ashraf Morningstar 
 * for learning and skill development. Original project concepts remains the intellectual 
 * property of their respective creators.
 */
/**
 * @fileoverview Rule to flag use of with statement
 * @author Nicholas C. Zakas
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

/** @type {import('../types').Rule.RuleModule} */
module.exports = {
	meta: {
		type: "suggestion",

		docs: {
			description: "Disallow `with` statements",
			recommended: true,
			url: "https://eslint.org/docs/latest/rules/no-with",
		},

		schema: [],

		messages: {
			unexpectedWith: "Unexpected use of 'with' statement.",
		},
	},

	create(context) {
		return {
			WithStatement(node) {
				context.report({ node, messageId: "unexpectedWith" });
			},
		};
	},
};
